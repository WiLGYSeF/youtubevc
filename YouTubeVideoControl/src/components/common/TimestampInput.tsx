import React, { ChangeEvent, KeyboardEvent } from 'react';

import secondsToTimestamp from '../../utils/secondsToTimestamp';
import timestampToSeconds from '../../utils/timestampToSeconds';
import trimstr from '../../utils/trimstr';
import useStatePropBacked from '../../utils/useStatePropBacked';

interface YtpcInputTimeProps {
  value?: string | number;
  setTime(seconds: number): void;
}

const TIME_PLACEHOLDER = '00:00';

function sanitizeTime(time: string): string {
  return trimstr(trimstr(time, ':'), '.').replace(/::+/, ':');
}

function strToSeconds(str: string): number {
  const sanitized = sanitizeTime(str);
  return sanitized ? timestampToSeconds(sanitized) : 0;
}

function getWidth(str: string): string {
  return `${Math.max(str.length, TIME_PLACEHOLDER.length)}ch`;
}

function YtpcInputTime(props: YtpcInputTimeProps) {
  const value = props.value ?? '';
  const [input, setInput] = useStatePropBacked(typeof value === 'number'
    ? secondsToTimestamp(value)
    : value);
  const [time, setTime] = useStatePropBacked(typeof value === 'number'
    ? value
    : strToSeconds(input));

  const onInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    const val = e.target.value;
    console.log(val);

    if (/^(\d?\d?:){0,3}\d?\d?(\.\d*)?$/.test(val)) {
      setInput(val);

      try {
        const seconds = strToSeconds(val);
        props.setTime(seconds);
        setTime(seconds);
      } catch (exc) {
        console.error(exc, val, sanitizeTime(val));
        props.setTime(-1);
        setTime(-1);
      }
    }
  };

  const updateInput = () => {
    setInput(secondsToTimestamp(time));
  };

  return (
    <span className="timestamp">
      <input
        style={{
          width: getWidth(input),
        }}
        value={input}
        placeholder={TIME_PLACEHOLDER}
        onChange={onInputChange}
        onBlur={updateInput}
        onKeyUp={(e: KeyboardEvent<HTMLInputElement>) => {
          if (e.key === 'Enter') {
            updateInput();
          }
        }}
      />
    </span>
  );
}

export default YtpcInputTime;
