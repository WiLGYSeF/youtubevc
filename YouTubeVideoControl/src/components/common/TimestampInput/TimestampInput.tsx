import React, {
  ChangeEvent, KeyboardEvent, useEffect,
} from 'react';

import { secondsToTimestamp, timestampToSeconds } from 'utils/timestr';
import trimstr from 'utils/trimstr';
import useStatePropBacked from 'utils/useStatePropBacked';

interface TimestampInputProps {
  defaultValue: number | string;
  onChange(seconds: number): void;
  setInput?(input: string): void;
}

const TIME_PLACEHOLDER = '00:00';

function sanitizeTime(time: string): string {
  return trimstr(time, ':', '.').replace(/::+/, ':');
}

function strToSeconds(str: string): number {
  const sanitized = sanitizeTime(str);
  return sanitized.length ? timestampToSeconds(sanitized) : 0;
}

function getWidth(str: string): string {
  return `${Math.max(str.length, TIME_PLACEHOLDER.length)}ch`;
}

function TimestampInput(props: TimestampInputProps) {
  const value = props.defaultValue ?? '';
  const getInput = () => (secondsToTimestamp(typeof value === 'number'
    ? value
    : strToSeconds(value)
  ));
  const getTime = () => (typeof value === 'number'
    ? value
    : strToSeconds(value)
  );

  const [input, setInput] = useStatePropBacked(getInput());
  const [time, setTime] = useStatePropBacked(getTime());

  useEffect(() => {
    if (props.setInput) {
      props.setInput(input);
    }
  }, []);

  useEffect(() => {
    const updatedInput = getInput();
    const updatedTime = getTime();

    if (props.setInput) {
      props.setInput(updatedInput);
    }
    setInput(updatedInput);
    props.onChange(updatedTime);
    setTime(updatedTime);
  }, [value]);

  const onInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    const val = e.target.value;

    if (/^(\d{0,2}:){0,3}\d{0,2}(\.\d{0,5})?$/.test(val)) {
      if (props.setInput) {
        props.setInput(val);
      }
      setInput(val);

      try {
        const seconds = strToSeconds(val);
        props.onChange(seconds);
        setTime(seconds);
      } catch (exc) {
        props.onChange(-1);
        setTime(-1);
      }
    } else {
      if (props.setInput) {
        props.setInput(input);
      }
      props.onChange(time);
    }
  };

  const updateInput = () => {
    const val = secondsToTimestamp(time);
    if (props.setInput) {
      props.setInput(val);
    }
    setInput(val);
  };

  return (
    <span className="timestamp">
      <input
        style={{
          width: getWidth(input),
        }}
        value={input}
        placeholder={TIME_PLACEHOLDER}
        onChange={onInputChange}
        onBlur={updateInput}
        onKeyUp={(e: KeyboardEvent<HTMLInputElement>) => {
          if (e.key === 'Enter') {
            updateInput();
          }
        }}
      />
    </span>
  );
}

export default TimestampInput;
