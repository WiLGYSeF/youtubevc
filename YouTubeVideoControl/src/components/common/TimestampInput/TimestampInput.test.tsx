import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import TimestampInput, { getInputs } from './TimestampInput';

describe('TimestampInput', () => {
  it('renders with string input', () => {
    const { container } = render(<TimestampInput
      defaultValue="27:51"
      onChange={() => {}}
    />);

    const input = container.getElementsByTagName('input')[0];
    expect(input.value).toEqual('27:51');
  });

  it('renders with number input', () => {
    const { container } = render(<TimestampInput
      defaultValue={3607}
      onChange={() => {}}
    />);

    const { input } = getInputs(container);
    expect(input.value).toEqual('01:00:07');
  });

  it.each([
    ['1:39', [1, 1, 60 + 3, 60 + 39], ['1', '1:', '1:3', '1:39', '01:39']],
    ['10:23:45', [
      1, 10, 10,
      10 * 60 + 2, 10 * 60 + 23, 10 * 60 + 23,
      10 * 60 * 60 + 23 * 60 + 4, 10 * 60 * 60 + 23 * 60 + 45,
    ], ['1', '10', '10:', '10:2', '10:23', '10:23:', '10:23:4', '10:23:45', '10:23:45']],
    ['0:71', [0, 0, 7, 71], ['0', '0:', '0:7', '0:71', '01:11']],
    ['00:12.3', [0, 0, 0, 1, 12, 12, 12.3], ['0', '00', '00:', '00:1', '00:12', '00:12.', '00:12.3', '00:12.3']],
    ['00:123', [0, 0, 0, 1, 12, 12], ['0', '00', '00:', '00:1', '00:12', '00:12', '00:12']],
  ])(
    'inputs "%s"',
    (inputStr: string, expected: number[], expectedStr: string[]) => {
      let testValue = 0;
      const testChangeFn = jest.fn();
      const testInputFn = jest.fn();

      const { container } = render(<TimestampInput
        defaultValue={0}
        onChange={(value: number) => {
          testValue = value;
          testChangeFn(value);
        }}
        setInput={(value: string) => {
          testInputFn(value);
        }}
      />);

      const { input } = getInputs(container);

      userEvent.clear(input);

      testChangeFn.mockClear();
      testInputFn.mockClear();

      for (let i = 0; i < inputStr.length; i += 1) {
        userEvent.keyboard(inputStr[i]);
        expect(testValue).toEqual(expected[i]);
      }

      fireEvent.blur(input);

      expect(testChangeFn.mock.calls.map((x) => x[0])).toEqual(expected);
      expect(testInputFn.mock.calls.map((x) => x[0])).toEqual(expectedStr);
    },
  );

  it('forces update on enter', () => {
    const testFn = jest.fn();

    const { container } = render(<TimestampInput
      defaultValue="0:71"
      onChange={() => {}}
      setInput={testFn}
    />);

    const { input } = getInputs(container);

    expect(input.value).toEqual('01:11');

    userEvent.clear(input);
    userEvent.type(input, '00:03{enter}');
    expect(input.value).toEqual('00:03');

    userEvent.clear(input);
    testFn.mockClear();
    userEvent.type(input, '15{enter}');
    expect(input.value).toEqual('00:15');
    expect(testFn.mock.calls.map((x) => x[0])).toEqual(['1', '15', '00:15']);
  });

  it('updates value on props change', () => {
    let value = 5;

    const { container, rerender } = render(<TimestampInput
      defaultValue={value}
      onChange={() => {}}
    />);

    const { input } = getInputs(container);

    expect(input.value).toEqual('00:05');

    value = 17;
    rerender(<TimestampInput
      defaultValue={value}
      onChange={() => {}}
    />);

    expect(input.value).toEqual('00:17');
  });
});
