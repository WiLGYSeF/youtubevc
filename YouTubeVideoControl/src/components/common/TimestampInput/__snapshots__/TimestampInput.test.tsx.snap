// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimestampInput renders with string input 1`] = `
<span
  className="timestamp"
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    placeholder="00:00"
    style={
      Object {
        "width": "5ch",
      }
    }
    value="27:51"
  />
</span>
`;
